pinot-plugins/pinot-stream-ingestion/pinot-kafka-2.0/src/main/java/org/apache/pinot/plugin/stream/kafka20/metadata/KafkaMessageMetadataFactory.java

```diff
@@ -0,0 +1,29 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.pinot.plugin.stream.kafka20.metadata;
+
+import org.apache.kafka.clients.consumer.ConsumerRecord;
+
+
+public class KafkaMessageMetadataFactory implements MessageMetadataFactory {
+  @Override
+  public KafkaMessageMetadata create(ConsumerRecord<?, ?> consumerRecord) {
+    return new KafkaMessageMetadata(consumerRecord.timestamp());
```

Comment:
curious, what happens if this hasn't been explicitly populated on the producer side? what's the default value?

Reply:
On the client side, the way this works is the programmer can provide a timestamp at `ProducerRecord` creation time. If this is not provided at creation time Kafka client will set the timestamp before sending the message to the broker. 
On the broker side, the timestamp that Kafka will attach to the message will depend on how the topic is configured. If the topic is configure to use "create time" the the broker uses the timestamp set in the client side (this is the default behaviour); if the topic is configured to use "log append time" the the broker set the message timestamp when the message is persisted.